Improve build process

the build process itself not absolutely sucking
90% of the build release time is waiting for flint to rebuild, which it shouldn't ever need to do without a new version (of flint) 
but docker is pain

the current build instructions doesn't work according to the instructions in the docker read me.

When running this command for this version `1.4.16 ` or the latest one `2.1.0.16`

```shell
docker build -f Dockerfile.source --build-arg GIT_COMMIT=$(git log -1 --format=%h) -t quilibrium -t quilibrium:1.4.16 .
```

This is the output error. 

```shell
 => ERROR [build-qclient 7/8] RUN GOOS=linux GOARCH=arm64 ./build.sh -o qclient                                                                                                               0.5s
 => CANCELED [build-node 3/3] RUN ./build.sh && cp node /usr/bin                                                                                                                              0.5s
------
 > [build-qclient 7/8] RUN GOOS=linux GOARCH=arm64 ./build.sh -o qclient:
0.122 +++++ dirname ./build.sh
0.123 ++++ realpath .
0.123 +++ dirname /opt/ceremonyclient/client
0.124 ++ cd /opt/ceremonyclient
0.124 ++ pwd
0.124 + ROOT_DIR=/opt/ceremonyclient
0.124 + CLIENT_DIR=/opt/ceremonyclient/client
0.124 + BINARIES_DIR=/opt/ceremonyclient/target/release
0.124 + pushd /opt/ceremonyclient/client
0.124 + export CGO_ENABLED=1
0.124 + CGO_ENABLED=1
0.124 ++ uname
0.125 + os_type=Linux
0.125 + case "$os_type" in
0.125 + export 'CGO_LDFLAGS=-L/usr/local/lib -lflint -lgmp -lmpfr -ldl -lm -L/opt/ceremonyclient/target/release -lstdc++ -lvdf -lchannel -lferret -lverenc -lbulletproofs -lbls48581 -lcrypto -lssl -static'
0.125 + CGO_LDFLAGS='-L/usr/local/lib -lflint -lgmp -lmpfr -ldl -lm -L/opt/ceremonyclient/target/release -lstdc++ -lvdf -lchannel -lferret -lverenc -lbulletproofs -lbls48581 -lcrypto -lssl -static'
0.125 + go build -ldflags '-linkmode '\''external'\''' -o qclient
0.487 cmd/crossMint.go:12:2: missing go.sum entry for module providing package github.com/cloudflare/circl/sign/ed448 (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.487 cmd/all.go:12:2: missing go.sum entry for module providing package github.com/iden3/go-iden3-crypto/poseidon (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.487 ../go-libp2p/core/crypto/key_to_stdlib.go:10:2: missing go.sum entry for module providing package github.com/decred/dcrd/dcrec/secp256k1/v4 (imported by github.com/libp2p/go-libp2p/core/crypto); to add:
0.487 	go get github.com/libp2p/go-libp2p/core/crypto@v0.41.1
0.487 ../go-libp2p/core/crypto/secp256k1.go:12:2: missing go.sum entry for module providing package github.com/decred/dcrd/dcrec/secp256k1/v4/ecdsa (imported by github.com/libp2p/go-libp2p/core/crypto); to add:
0.487 	go get github.com/libp2p/go-libp2p/core/crypto@v0.41.1
0.487 ../go-libp2p/core/crypto/pb/crypto.pb.go:10:2: missing go.sum entry for module providing package google.golang.org/protobuf/reflect/protoreflect (imported by source.quilibrium.com/quilibrium/monorepo/go-libp2p-blossomsub/pb); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/go-libp2p-blossomsub/pb@v0.0.0-00010101000000-000000000000
0.487 ../go-libp2p/core/crypto/pb/crypto.pb.go:11:2: missing go.sum entry for module providing package google.golang.org/protobuf/runtime/protoimpl (imported by source.quilibrium.com/quilibrium/monorepo/go-libp2p-blossomsub/pb); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/go-libp2p-blossomsub/pb@v0.0.0-00010101000000-000000000000
0.487 ../go-libp2p/core/crypto/key.go:15:2: missing go.sum entry for module providing package google.golang.org/protobuf/proto (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.487 ../go-libp2p/core/peer/peer.go:9:2: missing go.sum entry for module providing package github.com/ipfs/go-cid (imported by github.com/libp2p/go-libp2p/core/peer); to add:
0.487 	go get github.com/libp2p/go-libp2p/core/peer@v0.41.1
0.487 /root/go/pkg/mod/github.com/multiformats/go-multihash@v0.2.3/register/blake2/multihash_blake2.go:18:2: missing go.sum entry for module providing package golang.org/x/crypto/blake2b (imported by source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves@v0.0.0-00010101000000-000000000000
0.487 /root/go/pkg/mod/github.com/multiformats/go-multihash@v0.2.3/register/blake2/multihash_blake2.go:19:2: missing go.sum entry for module providing package golang.org/x/crypto/blake2s (imported by github.com/multiformats/go-multihash/register/blake2); to add:
0.487 	go get github.com/multiformats/go-multihash/register/blake2@v0.2.3
0.487 /root/go/pkg/mod/github.com/multiformats/go-multihash@v0.2.3/register/blake3/multihash_blake3.go:15:2: missing go.sum entry for module providing package lukechampine.com/blake3 (imported by github.com/multiformats/go-multihash/register/blake3); to add:
0.487 	go get github.com/multiformats/go-multihash/register/blake3@v0.2.3
0.487 /root/go/pkg/mod/github.com/multiformats/go-multihash@v0.2.3/register/sha3/multihash_sha3.go:19:2: missing go.sum entry for module providing package golang.org/x/crypto/sha3 (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.487 ../go-multiaddr/multiaddr.go:11:2: missing go.sum entry for module providing package golang.org/x/exp/slices (imported by github.com/multiformats/go-multiaddr); to add:
0.487 	go get github.com/multiformats/go-multiaddr@v0.16.1
0.487 ../go-libp2p/core/peer/peer.go:12:2: missing go.sum entry for module providing package github.com/multiformats/go-multicodec (imported by github.com/libp2p/go-libp2p/core/peer); to add:
0.487 	go get github.com/libp2p/go-libp2p/core/peer@v0.41.1
0.487 cmd/all.go:18:2: missing go.sum entry for module providing package google.golang.org/grpc (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.487 cmd/root.go:18:2: missing go.sum entry for module providing package google.golang.org/grpc/credentials (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.487 cmd/root.go:19:2: missing go.sum entry for module providing package google.golang.org/grpc/credentials/insecure (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.487 cmd/proverPause.go:15:2: missing go.sum entry for module providing package google.golang.org/protobuf/types/known/anypb (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.487 cmd/all.go:19:2: no required module provides package source.quilibrium.com/quilibrium/monorepo/node/config; to add it:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/config
0.487 cmd/all.go:20:2: no required module provides package source.quilibrium.com/quilibrium/monorepo/node/execution/intrinsics/token/application; to add it:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/execution/intrinsics/token/application
0.487 ../node/keys/file.go:23:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/config (imported by source.quilibrium.com/quilibrium/monorepo/node/keys); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/keys@v0.0.0-00010101000000-000000000000
0.487 ../nekryptology/pkg/core/curves/ed448_curve.go:16:2: missing go.sum entry for module providing package github.com/cloudflare/circl/ecc/goldilocks (imported by source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves@v0.0.0-00010101000000-000000000000
0.487 ../nekryptology/pkg/core/curves/ed448_curve.go:17:2: missing go.sum entry for module providing package github.com/cloudflare/circl/math/fp448 (imported by source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves@v0.0.0-00010101000000-000000000000
0.487 /root/go/pkg/mod/github.com/consensys/gnark-crypto@v0.5.3/ecc/bls12-377/fp/asm.go:22:8: missing go.sum entry for module providing package golang.org/x/sys/cpu (imported by github.com/consensys/gnark-crypto/ecc/bls12-377/fp); to add:
0.487 	go get github.com/consensys/gnark-crypto/ecc/bls12-377/fp@v0.5.3
0.487 ../nekryptology/internal/hash.go:14:2: missing go.sum entry for module providing package golang.org/x/crypto/hkdf (imported by github.com/libp2p/go-libp2p/config); to add:
0.487 	go get github.com/libp2p/go-libp2p/config@v0.41.1
0.487 ../node/keys/ed448_key.go:12:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/crypto (imported by source.quilibrium.com/quilibrium/monorepo/node/keys); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/keys@v0.0.0-00010101000000-000000000000
0.487 ../node/keys/file.go:25:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/keys (imported by source.quilibrium.com/quilibrium/monorepo/node/keys); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/keys@v0.0.0-00010101000000-000000000000
0.487 ../go-libp2p/core/protocol/switch.go:7:2: missing go.sum entry for module providing package github.com/multiformats/go-multistream (imported by github.com/libp2p/go-libp2p/core/protocol); to add:
0.487 	go get github.com/libp2p/go-libp2p/core/protocol@v0.41.1
0.487 ../go-libp2p/core/metrics/bandwidth.go:7:2: missing go.sum entry for module providing package github.com/libp2p/go-flow-metrics (imported by github.com/libp2p/go-libp2p/core/metrics); to add:
0.487 	go get github.com/libp2p/go-libp2p/core/metrics@v0.41.1
0.487 ../go-libp2p/p2p/metricshelper/registerer.go:6:2: missing go.sum entry for module providing package github.com/prometheus/client_golang/prometheus (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 ../go-libp2p/x/rate/limiter.go:13:2: missing go.sum entry for module providing package golang.org/x/time/rate (imported by github.com/libp2p/go-libp2p/p2p/transport/quicreuse); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/quicreuse@v0.41.1
0.487 ../go-libp2p/p2p/host/autorelay/relay_finder.go:13:2: missing go.sum entry for module providing package golang.org/x/sync/errgroup (imported by github.com/libp2p/go-libp2p/p2p/host/autorelay); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/host/autorelay@v0.41.1
0.487 ../go-libp2p/p2p/net/nat/internal/nat/upnp.go:15:2: missing go.sum entry for module providing package github.com/koron/go-ssdp (imported by github.com/libp2p/go-libp2p/p2p/net/nat/internal/nat); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/net/nat/internal/nat@v0.41.1
0.487 ../go-libp2p/p2p/net/nat/internal/nat/nat.go:15:2: missing go.sum entry for module providing package github.com/libp2p/go-netroute (imported by github.com/libp2p/go-libp2p-kad-dht); to add:
0.487 	go get github.com/libp2p/go-libp2p-kad-dht@v0.23.0
0.487 /root/go/pkg/mod/github.com/flynn/noise@v1.1.0/cipher_suite.go:15:2: missing go.sum entry for module providing package golang.org/x/crypto/chacha20poly1305 (imported by github.com/libp2p/go-libp2p/p2p/security/noise); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/security/noise@v0.41.1
0.487 /root/go/pkg/mod/github.com/flynn/noise@v1.1.0/cipher_suite.go:16:2: missing go.sum entry for module providing package golang.org/x/crypto/curve25519 (imported by github.com/flynn/noise); to add:
0.487 	go get github.com/flynn/noise@v1.1.0
0.487 /root/go/pkg/mod/github.com/davidlazar/go-crypto@v0.0.0-20200604182044-b73af7476f6c/salsa20/salsa20.go:7:2: missing go.sum entry for module providing package golang.org/x/crypto/salsa20/salsa (imported by github.com/davidlazar/go-crypto/salsa20); to add:
0.487 	go get github.com/davidlazar/go-crypto/salsa20@v0.0.0-20200604182044-b73af7476f6c
0.487 ../go-libp2p/p2p/transport/quicreuse/tracer.go:10:2: missing go.sum entry for module providing package github.com/klauspost/compress/zstd (imported by github.com/libp2p/go-libp2p/p2p/transport/quicreuse); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/quicreuse@v0.41.1
0.487 ../go-libp2p/p2p/transport/quicreuse/config.go:6:2: missing go.sum entry for module providing package github.com/quic-go/quic-go (imported by github.com/libp2p/go-libp2p/config); to add:
0.487 	go get github.com/libp2p/go-libp2p/config@v0.41.1
0.487 ../go-libp2p/p2p/transport/quicreuse/connmgr.go:21:2: missing go.sum entry for module providing package github.com/quic-go/quic-go/logging (imported by github.com/libp2p/go-libp2p/p2p/transport/quicreuse); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/quicreuse@v0.41.1
0.487 ../go-libp2p/p2p/transport/quicreuse/connmgr.go:22:2: missing go.sum entry for module providing package github.com/quic-go/quic-go/metrics (imported by github.com/libp2p/go-libp2p/p2p/transport/quicreuse); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/quicreuse@v0.41.1
0.487 ../go-libp2p/p2p/transport/quicreuse/tracer.go:14:2: missing go.sum entry for module providing package github.com/quic-go/quic-go/qlog (imported by github.com/libp2p/go-libp2p/p2p/transport/quicreuse); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/quicreuse@v0.41.1
0.487 ../go-libp2p/p2p/transport/webrtc/udpmux/mux.go:17:2: missing go.sum entry for module providing package github.com/pion/ice/v4 (imported by github.com/libp2p/go-libp2p/p2p/transport/webrtc/udpmux); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/webrtc/udpmux@v0.41.1
0.487 /root/go/pkg/mod/github.com/pion/stun@v0.6.1/client.go:19:2: missing go.sum entry for module providing package github.com/pion/dtls/v2 (imported by github.com/pion/stun); to add:
0.487 	go get github.com/pion/stun@v0.6.1
0.487 /root/go/pkg/mod/github.com/pion/stun@v0.6.1/client.go:20:2: missing go.sum entry for module providing package github.com/pion/transport/v2 (imported by github.com/pion/stun); to add:
0.487 	go get github.com/pion/stun@v0.6.1
0.487 /root/go/pkg/mod/github.com/pion/stun@v0.6.1/client.go:21:2: missing go.sum entry for module providing package github.com/pion/transport/v2/stdnet (imported by github.com/pion/stun); to add:
0.487 	go get github.com/pion/stun@v0.6.1
0.487 /root/go/pkg/mod/github.com/pion/stun@v0.6.1/xoraddr.go:13:2: missing go.sum entry for module providing package github.com/pion/transport/v2/utils/xor (imported by github.com/pion/stun); to add:
0.487 	go get github.com/pion/stun@v0.6.1
0.487 ../go-libp2p/p2p/transport/webrtc/connection.go:18:2: missing go.sum entry for module providing package github.com/pion/datachannel (imported by github.com/libp2p/go-libp2p/p2p/transport/webrtc); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/webrtc@v0.41.1
0.487 ../go-libp2p/p2p/transport/webrtc/logger.go:9:2: missing go.sum entry for module providing package github.com/pion/logging (imported by github.com/libp2p/go-libp2p/p2p/transport/webrtc); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/webrtc@v0.41.1
0.487 ../go-libp2p/p2p/transport/webrtc/connection.go:19:2: missing go.sum entry for module providing package github.com/pion/sctp (imported by github.com/libp2p/go-libp2p/p2p/transport/webrtc); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/webrtc@v0.41.1
0.487 ../go-libp2p/p2p/transport/webrtc/connection.go:20:2: missing go.sum entry for module providing package github.com/pion/webrtc/v4 (imported by github.com/libp2p/go-libp2p/p2p/transport/webrtc); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/webrtc@v0.41.1
0.487 ../go-libp2p/p2p/transport/webtransport/crypto.go:23:2: missing go.sum entry for module providing package github.com/quic-go/quic-go/http3 (imported by github.com/libp2p/go-libp2p/p2p/transport/webtransport); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/webtransport@v0.41.1
0.487 ../go-libp2p/p2p/transport/webtransport/conn.go:11:2: missing go.sum entry for module providing package github.com/quic-go/webtransport-go (imported by github.com/libp2p/go-libp2p/p2p/transport/webtransport); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/transport/webtransport@v0.41.1
0.487 ../go-libp2p/p2p/host/resource-manager/sys_unix.go:6:2: missing go.sum entry for module providing package golang.org/x/sys/unix (imported by github.com/libp2p/go-libp2p/p2p/host/resource-manager); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/host/resource-manager@v0.41.1
0.487 ../go-multiaddr-dns/resolve.go:8:2: missing go.sum entry for module providing package github.com/miekg/dns (imported by github.com/multiformats/go-multiaddr-dns); to add:
0.487 	go get github.com/multiformats/go-multiaddr-dns@v0.4.1
0.487 ../go-libp2p/config/config.go:51:2: missing go.sum entry for module providing package go.uber.org/fx (imported by github.com/libp2p/go-libp2p); to add:
0.487 	go get github.com/libp2p/go-libp2p@v0.41.1
0.487 ../go-libp2p/config/config.go:52:2: missing go.sum entry for module providing package go.uber.org/fx/fxevent (imported by github.com/libp2p/go-libp2p/config); to add:
0.487 	go get github.com/libp2p/go-libp2p/config@v0.41.1
0.487 ../go-libp2p/p2p/muxer/yamux/conn.go:8:2: missing go.sum entry for module providing package github.com/libp2p/go-yamux/v5 (imported by github.com/libp2p/go-libp2p/p2p/muxer/yamux); to add:
0.487 	go get github.com/libp2p/go-libp2p/p2p/muxer/yamux@v0.41.1
0.487 ../go-libp2p-kad-dht/dht.go:35:2: missing go.sum entry for module providing package github.com/ipfs/go-datastore (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.487 ../go-libp2p-kad-dht/internal/tracing.go:9:2: missing go.sum entry for module providing package go.opentelemetry.io/otel (imported by github.com/libp2p/go-libp2p-kad-dht/internal); to add:
0.487 	go get github.com/libp2p/go-libp2p-kad-dht/internal@v0.23.0
0.487 ../go-libp2p-kad-dht/internal/tracing.go:10:2: missing go.sum entry for module providing package go.opentelemetry.io/otel/attribute (imported by github.com/libp2p/go-libp2p-kad-dht); to add:
0.487 	go get github.com/libp2p/go-libp2p-kad-dht@v0.23.0
0.487 ../go-libp2p-kad-dht/internal/tracing.go:11:2: missing go.sum entry for module providing package go.opentelemetry.io/otel/trace (imported by github.com/libp2p/go-libp2p-kad-dht); to add:
0.487 	go get github.com/libp2p/go-libp2p-kad-dht@v0.23.0
0.487 ../go-libp2p-kad-dht/internal/config/config.go:9:2: missing go.sum entry for module providing package github.com/ipfs/go-datastore/sync (imported by github.com/libp2p/go-libp2p-kad-dht/internal/config); to add:
0.487 	go get github.com/libp2p/go-libp2p-kad-dht/internal/config@v0.23.0
0.487 ../go-libp2p-kad-dht/providers/providers_manager.go:14:2: missing go.sum entry for module providing package github.com/ipfs/go-datastore/autobatch (imported by github.com/libp2p/go-libp2p-kad-dht/providers); to add:
0.487 	go get github.com/libp2p/go-libp2p-kad-dht/providers@v0.23.0
0.487 ../go-libp2p-kad-dht/providers/providers_manager.go:15:2: missing go.sum entry for module providing package github.com/ipfs/go-datastore/query (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.487 ../go-libp2p-kad-dht/pb/protocol_messenger.go:15:2: missing go.sum entry for module providing package go.opentelemetry.io/otel/codes (imported by github.com/libp2p/go-libp2p-kad-dht/pb); to add:
0.487 	go get github.com/libp2p/go-libp2p-kad-dht/pb@v0.23.0
0.487 /root/go/pkg/mod/github.com/minio/sha256-simd@v1.0.1/cpuid_other.go:23:2: missing go.sum entry for module providing package github.com/klauspost/cpuid/v2 (imported by github.com/minio/sha256-simd); to add:
0.487 	go get github.com/minio/sha256-simd@v1.0.1
0.487 ../node/p2p/peer_authenticator.go:27:2: missing go.sum entry for module providing package google.golang.org/grpc/codes (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 ../node/p2p/blossomsub.go:47:2: missing go.sum entry for module providing package google.golang.org/grpc/peer (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 ../node/p2p/peer_authenticator.go:30:2: missing go.sum entry for module providing package google.golang.org/grpc/status (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 ../node/p2p/blossomsub.go:48:2: missing go.sum entry for module providing package google.golang.org/protobuf/types/known/wrapperspb (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 ../node/p2p/peer_info_manager.go:8:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/lifecycle (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 /root/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware/providers/prometheus@v1.0.1/server_metrics.go:10:2: missing go.sum entry for module providing package google.golang.org/grpc/reflection (imported by github.com/grpc-ecosystem/go-grpc-middleware/providers/prometheus); to add:
0.487 	go get github.com/grpc-ecosystem/go-grpc-middleware/providers/prometheus@v1.0.1
0.487 ../node/p2p/blossomsub.go:54:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/protobufs (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 ../node/p2p/blossomsub.go:55:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/channel (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 ../node/p2p/peer_authenticator.go:35:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/consensus (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 ../node/p2p/blossomsub.go:56:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/p2p (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 ../node/p2p/blossomsub.go:57:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/utils/p2p (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.487 cmd/all.go:22:2: no required module provides package source.quilibrium.com/quilibrium/monorepo/node/protobufs; to add it:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/protobufs
0.487 ../pebble/batch.go:19:2: missing go.sum entry for module providing package github.com/cockroachdb/errors (imported by github.com/cockroachdb/pebble); to add:
0.487 	go get github.com/cockroachdb/pebble@v1.1.5
0.487 ../pebble/checkpoint.go:11:2: missing go.sum entry for module providing package github.com/cockroachdb/errors/oserror (imported by github.com/cockroachdb/pebble); to add:
0.487 	go get github.com/cockroachdb/pebble@v1.1.5
0.487 ../pebble/internal/batchskl/skl.go:67:2: missing go.sum entry for module providing package golang.org/x/exp/rand (imported by github.com/cockroachdb/pebble/internal/batchskl); to add:
0.487 	go get github.com/cockroachdb/pebble/internal/batchskl@v1.1.5
0.487 ../pebble/sstable/compression_cgo.go:13:2: missing go.sum entry for module providing package github.com/DataDog/zstd (imported by github.com/cockroachdb/pebble/sstable); to add:
0.487 	go get github.com/cockroachdb/pebble/sstable@v1.1.5
0.487 ../pebble/sstable/reader.go:15:2: missing go.sum entry for module providing package github.com/cespare/xxhash/v2 (imported by github.com/cockroachdb/pebble/sstable); to add:
0.487 	go get github.com/cockroachdb/pebble/sstable@v1.1.5
0.487 ../pebble/sstable/compression.go:13:2: missing go.sum entry for module providing package github.com/golang/snappy (imported by github.com/cockroachdb/pebble/sstable); to add:
0.487 	go get github.com/cockroachdb/pebble/sstable@v1.1.5
0.487 ../node/store/clock.go:10:2: missing go.sum entry for module providing package github.com/cockroachdb/pebble/v2 (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.487 ../node/store/pebble.go:14:2: missing go.sum entry for module providing package github.com/cockroachdb/pebble/v2/vfs (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.487 ../node/store/consensus.go:11:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/consensus (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.487 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.488 ../node/store/consensus.go:12:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/consensus/models (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.488 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.488 ../node/store/hypergraph.go:16:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/hypergraph (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.488 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.488 ../node/store/hypergraph.go:19:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/hypergraph (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.488 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.488 ../node/store/clock.go:15:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/store (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.488 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.488 ../node/store/clock.go:16:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/tries (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.488 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
------
ERROR: failed to solve: process "/bin/sh -c GOOS=${TARGETOS} GOARCH=${TARGETARCH} ./build.sh -o qclient" did not complete successfully: exit code: 1

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/jofsmdlbyr603868xgxcmrxhw
```

> The current build process is basically: `task build_node_<linux|darwin>_<version>`.
> 
> A list of options are in the [task file](https://github.com/QuilibriumNetwork/monorepo/blob/develop/Taskfile.yaml) and the relevant README is in the [DOCKER-README](https://github.com/QuilibriumNetwork/monorepo/blob/develop/DOCKER-README.md) for using/installing `task`.
> 
> It's pretty straight-forward from there after getting `task` installed:
> 
> # list tasks
> task -l
> 
> # build the Linux ARM64 node binary
> task build_node_arm64_linux
> 
> # build the Linux ARM64 qclient binary
> task build_qclient_arm64_linux
> 
> # ... etc for the variations
> 
> The built binary is in the output/dist folder: `./node/build/<linux_amd64|linux_arm64|darwin_arm64>/` as `node`

I've tried to build it from source using the command

```shell
task build:source
```

but I get a similiar outcome as before

```shell
user@192 monorepo % task build:source
task: [build:source] docker build \
-f Dockerfile.source \
--build-arg NODE_VERSION=2.1.0 \
--build-arg GIT_REPO=git@github.com:QuilibriumNetwork/monorepo \
--build-arg GIT_BRANCH=develop \
--build-arg GIT_COMMIT=ab99f10 \
-t ${QUILIBRIUM_IMAGE_NAME:-quilibrium}:2.1.0-source \
-t ${QUILIBRIUM_IMAGE_NAME:-quilibrium}:source \
.

[+] Building 1.6s (49/53)                                                                                                                                                     docker:desktop-linux
 => [internal] load build definition from Dockerfile.source                                                                                                                                   0.0s
 => => transferring dockerfile: 5.64kB                                                                                                                                                        0.0s
 => resolve image config for docker-image://docker.io/docker/dockerfile:1.7-labs                                                                                                              0.4s
 => CACHED docker-image://docker.io/docker/dockerfile:1.7-labs@sha256:b99fecfe00268a8b556fad7d9c37ee25d716ae08a5d7320e6d51c4dd83246894                                                        0.0s
 => [internal] load metadata for docker.io/library/ubuntu:24.04                                                                                                                               0.3s
 => [internal] load build context                                                                                                                                                             0.1s
 => => transferring context: 248.76kB                                                                                                                                                         0.1s
 => [base 1/7] FROM docker.io/library/ubuntu:24.04@sha256:c35e29c9450151419d9448b0fd75374fec4fff364a27f176fb458d472dfc9e54                                                                    0.0s
 => CACHED [stage-9 2/6] RUN apt-get update && apt-get install libflint-dev -y                                                                                                                0.0s
 => CANCELED [stage-9 3/6] RUN apt-get update && apt-get install -y ca-certificates                                                                                                           0.7s
 => CACHED [base 2/7] RUN apt-get update && apt-get install -y     build-essential     curl     git     cmake     libgmp-dev     libmpfr-dev     libmpfr6     wget     m4     pkg-config      0.0s
 => CACHED [base 3/7] RUN apt update && apt install -y wget &&     ARCH=$(dpkg --print-architecture) &&     case ${ARCH} in         amd64) GOARCH=amd64 ;;         arm64) GOARCH=arm64 ;;     0.0s
 => CACHED [base 4/7] RUN git clone https://github.com/flintlib/flint.git &&     cd flint &&     git checkout flint-3.0 &&     ./bootstrap.sh &&     ./configure         --prefix=/usr/local  0.0s
 => CACHED [base 5/7] COPY docker/rustup-init.sh /opt/rustup-init.sh                                                                                                                          0.0s
 => CACHED [base 6/7] RUN /opt/rustup-init.sh -y --profile minimal                                                                                                                            0.0s
 => CACHED [base 7/7] RUN cargo install uniffi-bindgen-go --git https://github.com/NordSecurity/uniffi-bindgen-go --tag v0.4.0+v0.28.3                                                        0.0s
 => CACHED [build  1/27] RUN go install github.com/fullstorydev/grpcurl/cmd/grpcurl@latest                                                                                                    0.0s
 => CACHED [build  2/27] WORKDIR /opt/ceremonyclient                                                                                                                                          0.0s
 => CACHED [build  3/27] COPY --exclude=node     --exclude=client     --exclude=sidecar . .                                                                                                   0.0s
 => CACHED [build  4/27] RUN python emp-install.py --install --tool --ot                                                                                                                      0.0s
 => CACHED [build  5/27] RUN cd emp-tool && sed -i 's/add_library(${NAME} SHARED ${sources})/add_library(${NAME} STATIC ${sources})/g' CMakeLists.txt && mkdir build && cd build && cmake ..  0.0s
 => CACHED [build  6/27] RUN cd emp-ot && mkdir build && cd build && cmake .. -DCMAKE_INSTALL_PREFIX=/usr/local && cd .. && make && make install && cd ..                                     0.0s
 => CACHED [build  7/27] WORKDIR /opt/ceremonyclient/channel                                                                                                                                  0.0s
 => CACHED [build  8/27] RUN go mod download                                                                                                                                                  0.0s
 => CACHED [build  9/27] RUN ./generate.sh                                                                                                                                                    0.0s
 => CACHED [build 10/27] WORKDIR /opt/ceremonyclient/vdf                                                                                                                                      0.0s
 => CACHED [build 11/27] RUN go mod download                                                                                                                                                  0.0s
 => CACHED [build 12/27] RUN ./generate.sh                                                                                                                                                    0.0s
 => CACHED [build 13/27] WORKDIR /opt/ceremonyclient/ferret                                                                                                                                   0.0s
 => CACHED [build 14/27] RUN go mod download                                                                                                                                                  0.0s
 => CACHED [build 15/27] RUN ./generate.sh                                                                                                                                                    0.0s
 => CACHED [build 16/27] WORKDIR /opt/ceremonyclient/bls48581                                                                                                                                 0.0s
 => CACHED [build 17/27] RUN go mod download                                                                                                                                                  0.0s
 => CACHED [build 18/27] RUN ./generate.sh                                                                                                                                                    0.0s
 => CACHED [build 19/27] WORKDIR /opt/ceremonyclient/rpm                                                                                                                                      0.0s
 => CACHED [build 20/27] RUN go mod download                                                                                                                                                  0.0s
 => CACHED [build 21/27] RUN ./generate.sh                                                                                                                                                    0.0s
 => CACHED [build 22/27] WORKDIR /opt/ceremonyclient/verenc                                                                                                                                   0.0s
 => CACHED [build 23/27] RUN go mod download                                                                                                                                                  0.0s
 => CACHED [build 24/27] RUN ./generate.sh                                                                                                                                                    0.0s
 => CACHED [build 25/27] WORKDIR /opt/ceremonyclient/bulletproofs                                                                                                                             0.0s
 => CACHED [build 26/27] RUN go mod download                                                                                                                                                  0.0s
 => CACHED [build 27/27] RUN ./generate.sh                                                                                                                                                    0.0s
 => CACHED [build-qclient 1/8] COPY ./node /opt/ceremonyclient/node                                                                                                                           0.0s
 => CACHED [build-qclient 2/8] WORKDIR /opt/ceremonyclient/node                                                                                                                               0.0s
 => CACHED [build-qclient 3/8] RUN go mod download                                                                                                                                            0.0s
 => CACHED [build-qclient 4/8] COPY ./client /opt/ceremonyclient/client                                                                                                                       0.0s
 => CACHED [build-qclient 5/8] WORKDIR /opt/ceremonyclient/client                                                                                                                             0.0s
 => CACHED [build-qclient 6/8] RUN go mod download                                                                                                                                            0.0s
 => ERROR [build-qclient 7/8] RUN GOOS=linux GOARCH=arm64 ./build.sh -o qclient                                                                                                               0.6s
 => CANCELED [build-node 3/3] RUN ./build.sh && cp node /usr/bin                                                                                                                              0.6s
------
 > [build-qclient 7/8] RUN GOOS=linux GOARCH=arm64 ./build.sh -o qclient:
0.131 +++++ dirname ./build.sh
0.131 ++++ realpath .
0.132 +++ dirname /opt/ceremonyclient/client
0.133 ++ cd /opt/ceremonyclient
0.133 ++ pwd
0.133 + ROOT_DIR=/opt/ceremonyclient
0.133 + CLIENT_DIR=/opt/ceremonyclient/client
0.133 + BINARIES_DIR=/opt/ceremonyclient/target/release
0.133 + pushd /opt/ceremonyclient/client
0.133 + export CGO_ENABLED=1
0.133 + CGO_ENABLED=1
0.134 ++ uname
0.134 + os_type=Linux
0.134 + case "$os_type" in
0.134 + export 'CGO_LDFLAGS=-L/usr/local/lib -lflint -lgmp -lmpfr -ldl -lm -L/opt/ceremonyclient/target/release -lstdc++ -lvdf -lchannel -lferret -lverenc -lbulletproofs -lbls48581 -lcrypto -lssl -static'
0.134 + CGO_LDFLAGS='-L/usr/local/lib -lflint -lgmp -lmpfr -ldl -lm -L/opt/ceremonyclient/target/release -lstdc++ -lvdf -lchannel -lferret -lverenc -lbulletproofs -lbls48581 -lcrypto -lssl -static'
0.134 + go build -ldflags '-linkmode '\''external'\''' -o qclient
0.529 cmd/crossMint.go:12:2: missing go.sum entry for module providing package github.com/cloudflare/circl/sign/ed448 (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.529 cmd/all.go:12:2: missing go.sum entry for module providing package github.com/iden3/go-iden3-crypto/poseidon (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.529 ../go-libp2p/core/crypto/key_to_stdlib.go:10:2: missing go.sum entry for module providing package github.com/decred/dcrd/dcrec/secp256k1/v4 (imported by github.com/libp2p/go-libp2p/core/crypto); to add:
0.529 	go get github.com/libp2p/go-libp2p/core/crypto@v0.41.1
0.529 ../go-libp2p/core/crypto/secp256k1.go:12:2: missing go.sum entry for module providing package github.com/decred/dcrd/dcrec/secp256k1/v4/ecdsa (imported by github.com/libp2p/go-libp2p/core/crypto); to add:
0.529 	go get github.com/libp2p/go-libp2p/core/crypto@v0.41.1
0.529 ../go-libp2p/core/crypto/pb/crypto.pb.go:10:2: missing go.sum entry for module providing package google.golang.org/protobuf/reflect/protoreflect (imported by source.quilibrium.com/quilibrium/monorepo/go-libp2p-blossomsub/pb); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/go-libp2p-blossomsub/pb@v0.0.0-00010101000000-000000000000
0.529 ../go-libp2p/core/crypto/pb/crypto.pb.go:11:2: missing go.sum entry for module providing package google.golang.org/protobuf/runtime/protoimpl (imported by source.quilibrium.com/quilibrium/monorepo/go-libp2p-blossomsub/pb); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/go-libp2p-blossomsub/pb@v0.0.0-00010101000000-000000000000
0.529 ../go-libp2p/core/crypto/key.go:15:2: missing go.sum entry for module providing package google.golang.org/protobuf/proto (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.529 ../go-libp2p/core/peer/peer.go:9:2: missing go.sum entry for module providing package github.com/ipfs/go-cid (imported by github.com/libp2p/go-libp2p/core/peer); to add:
0.529 	go get github.com/libp2p/go-libp2p/core/peer@v0.41.1
0.529 /root/go/pkg/mod/github.com/multiformats/go-multihash@v0.2.3/register/blake2/multihash_blake2.go:18:2: missing go.sum entry for module providing package golang.org/x/crypto/blake2b (imported by source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves@v0.0.0-00010101000000-000000000000
0.529 /root/go/pkg/mod/github.com/multiformats/go-multihash@v0.2.3/register/blake2/multihash_blake2.go:19:2: missing go.sum entry for module providing package golang.org/x/crypto/blake2s (imported by github.com/multiformats/go-multihash/register/blake2); to add:
0.529 	go get github.com/multiformats/go-multihash/register/blake2@v0.2.3
0.529 /root/go/pkg/mod/github.com/multiformats/go-multihash@v0.2.3/register/blake3/multihash_blake3.go:15:2: missing go.sum entry for module providing package lukechampine.com/blake3 (imported by github.com/multiformats/go-multihash/register/blake3); to add:
0.529 	go get github.com/multiformats/go-multihash/register/blake3@v0.2.3
0.529 /root/go/pkg/mod/github.com/multiformats/go-multihash@v0.2.3/register/sha3/multihash_sha3.go:19:2: missing go.sum entry for module providing package golang.org/x/crypto/sha3 (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.529 ../go-multiaddr/multiaddr.go:11:2: missing go.sum entry for module providing package golang.org/x/exp/slices (imported by github.com/multiformats/go-multiaddr); to add:
0.529 	go get github.com/multiformats/go-multiaddr@v0.16.1
0.529 ../go-libp2p/core/peer/peer.go:12:2: missing go.sum entry for module providing package github.com/multiformats/go-multicodec (imported by github.com/libp2p/go-libp2p/core/peer); to add:
0.529 	go get github.com/libp2p/go-libp2p/core/peer@v0.41.1
0.529 cmd/all.go:18:2: missing go.sum entry for module providing package google.golang.org/grpc (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.529 cmd/root.go:18:2: missing go.sum entry for module providing package google.golang.org/grpc/credentials (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.529 cmd/root.go:19:2: missing go.sum entry for module providing package google.golang.org/grpc/credentials/insecure (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.529 cmd/proverPause.go:15:2: missing go.sum entry for module providing package google.golang.org/protobuf/types/known/anypb (imported by source.quilibrium.com/quilibrium/monorepo/client/cmd); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/client/cmd
0.529 cmd/all.go:19:2: no required module provides package source.quilibrium.com/quilibrium/monorepo/node/config; to add it:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/config
0.529 cmd/all.go:20:2: no required module provides package source.quilibrium.com/quilibrium/monorepo/node/execution/intrinsics/token/application; to add it:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/execution/intrinsics/token/application
0.529 ../node/keys/file.go:23:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/config (imported by source.quilibrium.com/quilibrium/monorepo/node/keys); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/keys@v0.0.0-00010101000000-000000000000
0.529 ../nekryptology/pkg/core/curves/ed448_curve.go:16:2: missing go.sum entry for module providing package github.com/cloudflare/circl/ecc/goldilocks (imported by source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves@v0.0.0-00010101000000-000000000000
0.529 ../nekryptology/pkg/core/curves/ed448_curve.go:17:2: missing go.sum entry for module providing package github.com/cloudflare/circl/math/fp448 (imported by source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves@v0.0.0-00010101000000-000000000000
0.529 /root/go/pkg/mod/github.com/consensys/gnark-crypto@v0.5.3/ecc/bls12-377/fp/asm.go:22:8: missing go.sum entry for module providing package golang.org/x/sys/cpu (imported by github.com/consensys/gnark-crypto/ecc/bls12-377/fp); to add:
0.529 	go get github.com/consensys/gnark-crypto/ecc/bls12-377/fp@v0.5.3
0.529 ../nekryptology/internal/hash.go:14:2: missing go.sum entry for module providing package golang.org/x/crypto/hkdf (imported by github.com/libp2p/go-libp2p/config); to add:
0.529 	go get github.com/libp2p/go-libp2p/config@v0.41.1
0.529 ../node/keys/ed448_key.go:12:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/crypto (imported by source.quilibrium.com/quilibrium/monorepo/node/keys); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/keys@v0.0.0-00010101000000-000000000000
0.529 ../node/keys/file.go:25:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/keys (imported by source.quilibrium.com/quilibrium/monorepo/node/keys); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/keys@v0.0.0-00010101000000-000000000000
0.529 ../go-libp2p/core/protocol/switch.go:7:2: missing go.sum entry for module providing package github.com/multiformats/go-multistream (imported by github.com/libp2p/go-libp2p/core/protocol); to add:
0.529 	go get github.com/libp2p/go-libp2p/core/protocol@v0.41.1
0.529 ../go-libp2p/core/metrics/bandwidth.go:7:2: missing go.sum entry for module providing package github.com/libp2p/go-flow-metrics (imported by github.com/libp2p/go-libp2p/core/metrics); to add:
0.529 	go get github.com/libp2p/go-libp2p/core/metrics@v0.41.1
0.529 ../go-libp2p/p2p/metricshelper/registerer.go:6:2: missing go.sum entry for module providing package github.com/prometheus/client_golang/prometheus (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 ../go-libp2p/x/rate/limiter.go:13:2: missing go.sum entry for module providing package golang.org/x/time/rate (imported by github.com/libp2p/go-libp2p/p2p/transport/quicreuse); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/quicreuse@v0.41.1
0.529 ../go-libp2p/p2p/host/autorelay/relay_finder.go:13:2: missing go.sum entry for module providing package golang.org/x/sync/errgroup (imported by github.com/libp2p/go-libp2p/p2p/host/autorelay); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/host/autorelay@v0.41.1
0.529 ../go-libp2p/p2p/net/nat/internal/nat/upnp.go:15:2: missing go.sum entry for module providing package github.com/koron/go-ssdp (imported by github.com/libp2p/go-libp2p/p2p/net/nat/internal/nat); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/net/nat/internal/nat@v0.41.1
0.529 ../go-libp2p/p2p/net/nat/internal/nat/nat.go:15:2: missing go.sum entry for module providing package github.com/libp2p/go-netroute (imported by github.com/libp2p/go-libp2p-kad-dht); to add:
0.529 	go get github.com/libp2p/go-libp2p-kad-dht@v0.23.0
0.529 /root/go/pkg/mod/github.com/flynn/noise@v1.1.0/cipher_suite.go:15:2: missing go.sum entry for module providing package golang.org/x/crypto/chacha20poly1305 (imported by github.com/libp2p/go-libp2p/p2p/security/noise); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/security/noise@v0.41.1
0.529 /root/go/pkg/mod/github.com/flynn/noise@v1.1.0/cipher_suite.go:16:2: missing go.sum entry for module providing package golang.org/x/crypto/curve25519 (imported by github.com/flynn/noise); to add:
0.529 	go get github.com/flynn/noise@v1.1.0
0.529 /root/go/pkg/mod/github.com/davidlazar/go-crypto@v0.0.0-20200604182044-b73af7476f6c/salsa20/salsa20.go:7:2: missing go.sum entry for module providing package golang.org/x/crypto/salsa20/salsa (imported by github.com/davidlazar/go-crypto/salsa20); to add:
0.529 	go get github.com/davidlazar/go-crypto/salsa20@v0.0.0-20200604182044-b73af7476f6c
0.529 ../go-libp2p/p2p/transport/quicreuse/tracer.go:10:2: missing go.sum entry for module providing package github.com/klauspost/compress/zstd (imported by github.com/libp2p/go-libp2p/p2p/transport/quicreuse); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/quicreuse@v0.41.1
0.529 ../go-libp2p/p2p/transport/quicreuse/config.go:6:2: missing go.sum entry for module providing package github.com/quic-go/quic-go (imported by github.com/libp2p/go-libp2p/config); to add:
0.529 	go get github.com/libp2p/go-libp2p/config@v0.41.1
0.529 ../go-libp2p/p2p/transport/quicreuse/connmgr.go:21:2: missing go.sum entry for module providing package github.com/quic-go/quic-go/logging (imported by github.com/libp2p/go-libp2p/p2p/transport/quicreuse); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/quicreuse@v0.41.1
0.529 ../go-libp2p/p2p/transport/quicreuse/connmgr.go:22:2: missing go.sum entry for module providing package github.com/quic-go/quic-go/metrics (imported by github.com/libp2p/go-libp2p/p2p/transport/quicreuse); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/quicreuse@v0.41.1
0.529 ../go-libp2p/p2p/transport/quicreuse/tracer.go:14:2: missing go.sum entry for module providing package github.com/quic-go/quic-go/qlog (imported by github.com/libp2p/go-libp2p/p2p/transport/quicreuse); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/quicreuse@v0.41.1
0.529 ../go-libp2p/p2p/transport/webrtc/udpmux/mux.go:17:2: missing go.sum entry for module providing package github.com/pion/ice/v4 (imported by github.com/libp2p/go-libp2p/p2p/transport/webrtc/udpmux); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/webrtc/udpmux@v0.41.1
0.529 /root/go/pkg/mod/github.com/pion/stun@v0.6.1/client.go:19:2: missing go.sum entry for module providing package github.com/pion/dtls/v2 (imported by github.com/pion/stun); to add:
0.529 	go get github.com/pion/stun@v0.6.1
0.529 /root/go/pkg/mod/github.com/pion/stun@v0.6.1/client.go:20:2: missing go.sum entry for module providing package github.com/pion/transport/v2 (imported by github.com/pion/stun); to add:
0.529 	go get github.com/pion/stun@v0.6.1
0.529 /root/go/pkg/mod/github.com/pion/stun@v0.6.1/client.go:21:2: missing go.sum entry for module providing package github.com/pion/transport/v2/stdnet (imported by github.com/pion/stun); to add:
0.529 	go get github.com/pion/stun@v0.6.1
0.529 /root/go/pkg/mod/github.com/pion/stun@v0.6.1/xoraddr.go:13:2: missing go.sum entry for module providing package github.com/pion/transport/v2/utils/xor (imported by github.com/pion/stun); to add:
0.529 	go get github.com/pion/stun@v0.6.1
0.529 ../go-libp2p/p2p/transport/webrtc/connection.go:18:2: missing go.sum entry for module providing package github.com/pion/datachannel (imported by github.com/libp2p/go-libp2p/p2p/transport/webrtc); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/webrtc@v0.41.1
0.529 ../go-libp2p/p2p/transport/webrtc/logger.go:9:2: missing go.sum entry for module providing package github.com/pion/logging (imported by github.com/libp2p/go-libp2p/p2p/transport/webrtc); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/webrtc@v0.41.1
0.529 ../go-libp2p/p2p/transport/webrtc/connection.go:19:2: missing go.sum entry for module providing package github.com/pion/sctp (imported by github.com/libp2p/go-libp2p/p2p/transport/webrtc); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/webrtc@v0.41.1
0.529 ../go-libp2p/p2p/transport/webrtc/connection.go:20:2: missing go.sum entry for module providing package github.com/pion/webrtc/v4 (imported by github.com/libp2p/go-libp2p/p2p/transport/webrtc); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/webrtc@v0.41.1
0.529 ../go-libp2p/p2p/transport/webtransport/crypto.go:23:2: missing go.sum entry for module providing package github.com/quic-go/quic-go/http3 (imported by github.com/libp2p/go-libp2p/p2p/transport/webtransport); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/webtransport@v0.41.1
0.529 ../go-libp2p/p2p/transport/webtransport/conn.go:11:2: missing go.sum entry for module providing package github.com/quic-go/webtransport-go (imported by github.com/libp2p/go-libp2p/p2p/transport/webtransport); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/transport/webtransport@v0.41.1
0.529 ../go-libp2p/p2p/host/resource-manager/sys_unix.go:6:2: missing go.sum entry for module providing package golang.org/x/sys/unix (imported by github.com/libp2p/go-libp2p/p2p/host/resource-manager); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/host/resource-manager@v0.41.1
0.529 ../go-multiaddr-dns/resolve.go:8:2: missing go.sum entry for module providing package github.com/miekg/dns (imported by github.com/multiformats/go-multiaddr-dns); to add:
0.529 	go get github.com/multiformats/go-multiaddr-dns@v0.4.1
0.529 ../go-libp2p/config/config.go:51:2: missing go.sum entry for module providing package go.uber.org/fx (imported by github.com/libp2p/go-libp2p); to add:
0.529 	go get github.com/libp2p/go-libp2p@v0.41.1
0.529 ../go-libp2p/config/config.go:52:2: missing go.sum entry for module providing package go.uber.org/fx/fxevent (imported by github.com/libp2p/go-libp2p/config); to add:
0.529 	go get github.com/libp2p/go-libp2p/config@v0.41.1
0.529 ../go-libp2p/p2p/muxer/yamux/conn.go:8:2: missing go.sum entry for module providing package github.com/libp2p/go-yamux/v5 (imported by github.com/libp2p/go-libp2p/p2p/muxer/yamux); to add:
0.529 	go get github.com/libp2p/go-libp2p/p2p/muxer/yamux@v0.41.1
0.529 ../go-libp2p-kad-dht/dht.go:35:2: missing go.sum entry for module providing package github.com/ipfs/go-datastore (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.529 ../go-libp2p-kad-dht/internal/tracing.go:9:2: missing go.sum entry for module providing package go.opentelemetry.io/otel (imported by github.com/libp2p/go-libp2p-kad-dht/internal); to add:
0.529 	go get github.com/libp2p/go-libp2p-kad-dht/internal@v0.23.0
0.529 ../go-libp2p-kad-dht/internal/tracing.go:10:2: missing go.sum entry for module providing package go.opentelemetry.io/otel/attribute (imported by github.com/libp2p/go-libp2p-kad-dht); to add:
0.529 	go get github.com/libp2p/go-libp2p-kad-dht@v0.23.0
0.529 ../go-libp2p-kad-dht/internal/tracing.go:11:2: missing go.sum entry for module providing package go.opentelemetry.io/otel/trace (imported by github.com/libp2p/go-libp2p-kad-dht); to add:
0.529 	go get github.com/libp2p/go-libp2p-kad-dht@v0.23.0
0.529 ../go-libp2p-kad-dht/internal/config/config.go:9:2: missing go.sum entry for module providing package github.com/ipfs/go-datastore/sync (imported by github.com/libp2p/go-libp2p-kad-dht/internal/config); to add:
0.529 	go get github.com/libp2p/go-libp2p-kad-dht/internal/config@v0.23.0
0.529 ../go-libp2p-kad-dht/providers/providers_manager.go:14:2: missing go.sum entry for module providing package github.com/ipfs/go-datastore/autobatch (imported by github.com/libp2p/go-libp2p-kad-dht/providers); to add:
0.529 	go get github.com/libp2p/go-libp2p-kad-dht/providers@v0.23.0
0.529 ../go-libp2p-kad-dht/providers/providers_manager.go:15:2: missing go.sum entry for module providing package github.com/ipfs/go-datastore/query (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.529 ../go-libp2p-kad-dht/pb/protocol_messenger.go:15:2: missing go.sum entry for module providing package go.opentelemetry.io/otel/codes (imported by github.com/libp2p/go-libp2p-kad-dht/pb); to add:
0.529 	go get github.com/libp2p/go-libp2p-kad-dht/pb@v0.23.0
0.529 /root/go/pkg/mod/github.com/minio/sha256-simd@v1.0.1/cpuid_other.go:23:2: missing go.sum entry for module providing package github.com/klauspost/cpuid/v2 (imported by github.com/minio/sha256-simd); to add:
0.529 	go get github.com/minio/sha256-simd@v1.0.1
0.529 ../node/p2p/peer_authenticator.go:27:2: missing go.sum entry for module providing package google.golang.org/grpc/codes (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 ../node/p2p/blossomsub.go:47:2: missing go.sum entry for module providing package google.golang.org/grpc/peer (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 ../node/p2p/peer_authenticator.go:30:2: missing go.sum entry for module providing package google.golang.org/grpc/status (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 ../node/p2p/blossomsub.go:48:2: missing go.sum entry for module providing package google.golang.org/protobuf/types/known/wrapperspb (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 ../node/p2p/peer_info_manager.go:8:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/lifecycle (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 /root/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware/providers/prometheus@v1.0.1/server_metrics.go:10:2: missing go.sum entry for module providing package google.golang.org/grpc/reflection (imported by github.com/grpc-ecosystem/go-grpc-middleware/providers/prometheus); to add:
0.529 	go get github.com/grpc-ecosystem/go-grpc-middleware/providers/prometheus@v1.0.1
0.529 ../node/p2p/blossomsub.go:54:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/protobufs (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 ../node/p2p/blossomsub.go:55:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/channel (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 ../node/p2p/peer_authenticator.go:35:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/consensus (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 ../node/p2p/blossomsub.go:56:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/p2p (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 ../node/p2p/blossomsub.go:57:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/utils/p2p (imported by source.quilibrium.com/quilibrium/monorepo/node/p2p); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/p2p@v0.0.0-00010101000000-000000000000
0.529 cmd/all.go:22:2: no required module provides package source.quilibrium.com/quilibrium/monorepo/node/protobufs; to add it:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/protobufs
0.529 ../pebble/batch.go:19:2: missing go.sum entry for module providing package github.com/cockroachdb/errors (imported by github.com/cockroachdb/pebble); to add:
0.529 	go get github.com/cockroachdb/pebble@v1.1.5
0.529 ../pebble/checkpoint.go:11:2: missing go.sum entry for module providing package github.com/cockroachdb/errors/oserror (imported by github.com/cockroachdb/pebble); to add:
0.529 	go get github.com/cockroachdb/pebble@v1.1.5
0.529 ../pebble/internal/batchskl/skl.go:67:2: missing go.sum entry for module providing package golang.org/x/exp/rand (imported by github.com/cockroachdb/pebble/internal/batchskl); to add:
0.529 	go get github.com/cockroachdb/pebble/internal/batchskl@v1.1.5
0.529 ../pebble/sstable/compression_cgo.go:13:2: missing go.sum entry for module providing package github.com/DataDog/zstd (imported by github.com/cockroachdb/pebble/sstable); to add:
0.529 	go get github.com/cockroachdb/pebble/sstable@v1.1.5
0.529 ../pebble/sstable/reader.go:15:2: missing go.sum entry for module providing package github.com/cespare/xxhash/v2 (imported by github.com/cockroachdb/pebble/sstable); to add:
0.529 	go get github.com/cockroachdb/pebble/sstable@v1.1.5
0.529 ../pebble/sstable/compression.go:13:2: missing go.sum entry for module providing package github.com/golang/snappy (imported by github.com/cockroachdb/pebble/sstable); to add:
0.529 	go get github.com/cockroachdb/pebble/sstable@v1.1.5
0.529 ../node/store/clock.go:10:2: missing go.sum entry for module providing package github.com/cockroachdb/pebble/v2 (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.529 ../node/store/pebble.go:14:2: missing go.sum entry for module providing package github.com/cockroachdb/pebble/v2/vfs (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.529 ../node/store/consensus.go:11:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/consensus (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.529 ../node/store/consensus.go:12:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/consensus/models (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.529 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.530 ../node/store/hypergraph.go:16:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/hypergraph (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.530 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.530 ../node/store/hypergraph.go:19:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/hypergraph (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.530 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.530 ../node/store/clock.go:15:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/store (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.530 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
0.530 ../node/store/clock.go:16:2: missing go.sum entry for module providing package source.quilibrium.com/quilibrium/monorepo/types/tries (imported by source.quilibrium.com/quilibrium/monorepo/node/store); to add:
0.530 	go get source.quilibrium.com/quilibrium/monorepo/node/store@v0.0.0-00010101000000-000000000000
------
ERROR: failed to solve: process "/bin/sh -c GOOS=${TARGETOS} GOARCH=${TARGETARCH} ./build.sh -o qclient" did not complete successfully: exit code: 1

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/mbxdek3hv5ahglxzjxeh33scm
task: Failed to run task "build:source": exit status 1
user@192 monorepo % 

```

--- a/Dockerfile.source
+++ b/Dockerfile.source
@@ -1,12 +1,13 @@
 # syntax=docker.io/docker/dockerfile:1.7-labs
 FROM --platform=${TARGETPLATFORM} ubuntu:24.04 AS base
 
+# Common environment variables
 ENV PATH="${PATH}:/root/.cargo/bin/"
-
 ARG TARGETOS
 ARG TARGETARCH
+ARG GO_VERSION=1.23.5
 
-# Install GMP 6.2 (6.3 which MacOS is using only available on Debian unstable)
+# Install basics
 RUN apt-get update && apt-get install -y \
     build-essential \
     curl \
@@ -27,15 +28,64 @@ RUN apt-get update && apt-get install -y \
     libssl-dev \
     python3 \
     python-is-python3 \
+    libflint-dev \
+    ca-certificates \
     && rm -rf /var/lib/apt/lists/*
 
-ARG GO_VERSION=1.23.5
+# -----------------------------------------------------------------------------
+# Stage: flint-builder
+# Purpose: Build FLINT library (slowest component)
+# -----------------------------------------------------------------------------
+FROM base AS flint-builder
+
+RUN git clone --branch flint-3.0 --depth 1 https://github.com/flintlib/flint.git && \
+    cd flint && \
+    ./bootstrap.sh && \
+    ./configure \
+    --prefix=/usr/local \
+    --with-gmp=/usr/local \
+    --with-mpfr=/usr/local \
+    --enable-static \
+    --disable-shared \
+    CFLAGS="-O3" && \
+    make -j$(nproc) && \
+    make install
+
+# -----------------------------------------------------------------------------
+# Stage: emp-builder
+# Purpose: Build EMP-toolkit libraries
+# -----------------------------------------------------------------------------
+FROM base AS emp-builder
+
+WORKDIR /opt/ceremonyclient
+# Copy the necessary script and source directories
+COPY emp-install.py .
+COPY emp-tool emp-tool
+COPY emp-ot emp-ot
+
+RUN python emp-install.py --install --tool --ot
+
+# Fix emp-tool to be static and install
+RUN cd emp-tool && sed -i 's/add_library(${NAME} SHARED ${sources})/add_library(${NAME} STATIC ${sources})/g' CMakeLists.txt && mkdir build && cd build && cmake .. -DCMAKE_INSTALL_PREFIX=/usr/local && cd .. && make -j$(nproc) && make install
+
+# Install emp-ot
+RUN cd emp-ot && mkdir build && cd build && cmake .. -DCMAKE_INSTALL_PREFIX=/usr/local && cd .. && make -j$(nproc) && make install
+
+# -----------------------------------------------------------------------------
+# Stage: go-base
+# Purpose: Install Go
+# -----------------------------------------------------------------------------
+FROM base AS go-base
+
+ARG TARGETARCH
+ARG GO_VERSION
+
 RUN apt update && apt install -y wget && \
     ARCH=$(dpkg --print-architecture) && \
     case ${ARCH} in \
-        amd64) GOARCH=amd64 ;; \
-        arm64) GOARCH=arm64 ;; \
-        *) echo "Unsupported architecture: ${ARCH}" && exit 1 ;; \
+    amd64) GOARCH=amd64 ;; \
+    arm64) GOARCH=arm64 ;; \
+    *) echo "Unsupported architecture: ${ARCH}" && exit 1 ;; \
     esac && \
     wget https://go.dev/dl/go${GO_VERSION}.linux-${GOARCH}.tar.gz && \
     rm -rf /usr/local/go && \
@@ -43,139 +93,186 @@ RUN apt update && apt install -y wget && \
     rm go${GO_VERSION}.linux-${GOARCH}.tar.gz
 
 ENV PATH=$PATH:/usr/local/go/bin
+ENV GOPROXY=direct
 
-RUN git clone https://github.com/flintlib/flint.git && \
-    cd flint && \
-    git checkout flint-3.0 && \
-    ./bootstrap.sh && \
-    ./configure \
-        --prefix=/usr/local \
-        --with-gmp=/usr/local \
-        --with-mpfr=/usr/local \
-        --enable-static \
-        --disable-shared \
-        CFLAGS="-O3" && \
-    make && \
-    make install && \
-    cd .. && \
-    rm -rf flint
+# -----------------------------------------------------------------------------
+# Stage: rust-base
+# Purpose: Install Rust and bindgen tools
+# -----------------------------------------------------------------------------
+FROM base AS rust-base
 
 COPY docker/rustup-init.sh /opt/rustup-init.sh
-
 RUN /opt/rustup-init.sh -y --profile minimal
 
-# Install uniffi-bindgen-go
 RUN cargo install uniffi-bindgen-go --git https://github.com/NordSecurity/uniffi-bindgen-go --tag v0.4.0+v0.28.3
 
-FROM base AS build
+# -----------------------------------------------------------------------------
+# Stage: libs-context
+# Purpose: Aggregate built libraries (Flint + EMP)
+# -----------------------------------------------------------------------------
+FROM base AS libs-context
+COPY --from=flint-builder /usr/local /usr/local
+COPY --from=emp-builder /usr/local /usr/local
+
+# -----------------------------------------------------------------------------
+# Stage: common-context
+# Purpose: Base for generation steps, includes libs, go, and rust
+# -----------------------------------------------------------------------------
+FROM libs-context AS common-context
+
+COPY --from=go-base /usr/local/go /usr/local/go
+ENV PATH=$PATH:/usr/local/go/bin
+ENV GOPROXY=https://proxy.golang.org,direct
+
+COPY --from=rust-base /root/.cargo /root/.cargo
+COPY --from=rust-base /root/.rustup /root/.rustup
+ENV PATH="${PATH}:/root/.cargo/bin/"
 
 ENV GOEXPERIMENT=arenas
 ENV QUILIBRIUM_SIGNATURE_CHECK=false
 
-# Install grpcurl before building the node and client 
-# as to avoid needing to redo it on rebuilds
+# Install grpcurl (common tool)
 RUN go install github.com/fullstorydev/grpcurl/cmd/grpcurl@latest
 
 WORKDIR /opt/ceremonyclient
 
-# Copy everything except node and client so as to avoid 
-# invalidating the cache at this point on client or node rebuilds
-
+# Copy source needed for generation (excluding node/client/sidecar handled in builders)
 COPY --exclude=node \
     --exclude=client \
     --exclude=sidecar . .
 
-RUN python emp-install.py --install --tool --ot
+# -----------------------------------------------------------------------------
+# Parallel Generation Stages
+# -----------------------------------------------------------------------------
 
-RUN cd emp-tool && sed -i 's/add_library(${NAME} SHARED ${sources})/add_library(${NAME} STATIC ${sources})/g' CMakeLists.txt && mkdir build && cd build && cmake .. -DCMAKE_INSTALL_PREFIX=/usr/local && cd .. && make && make install && cd ..
-
-RUN cd emp-ot && mkdir build && cd build && cmake .. -DCMAKE_INSTALL_PREFIX=/usr/local && cd .. && make && make install && cd ..
-
-## Generate Rust bindings for channel
+FROM common-context AS gen-channel
 WORKDIR /opt/ceremonyclient/channel
-
 RUN go mod download
-
 RUN ./generate.sh
 
-
-## Generate Rust bindings for VDF
+FROM common-context AS gen-vdf
 WORKDIR /opt/ceremonyclient/vdf
-
 RUN go mod download
-
 RUN ./generate.sh
 
-
-## Generate Rust bindings for Ferret
+FROM common-context AS gen-ferret
 WORKDIR /opt/ceremonyclient/ferret
-
 RUN go mod download
-
 RUN ./generate.sh
 
-## Generate Rust bindings for BLS48581
+FROM common-context AS gen-bls48581
 WORKDIR /opt/ceremonyclient/bls48581
-
 RUN go mod download
-
 RUN ./generate.sh
 
-## Generate Rust bindings for RPM
+FROM common-context AS gen-rpm
 WORKDIR /opt/ceremonyclient/rpm
-
 RUN go mod download
-
 RUN ./generate.sh
 
-## Generate Rust bindings for VerEnc
+FROM common-context AS gen-verenc
 WORKDIR /opt/ceremonyclient/verenc
-
 RUN go mod download
-
 RUN ./generate.sh
 
-## Generate Rust bindings for Bulletproofs
+FROM common-context AS gen-bulletproofs
 WORKDIR /opt/ceremonyclient/bulletproofs
-
 RUN go mod download
-
 RUN ./generate.sh
 
-FROM build AS build-node
+# -----------------------------------------------------------------------------
+# Stage: build-context
+# Purpose: Aggregate all generated bindings and static libraries
+# -----------------------------------------------------------------------------
+FROM common-context AS build-context
 
-# Build and install the node
+# Copy generated artifacts back
+COPY --from=gen-channel /opt/ceremonyclient/channel /opt/ceremonyclient/channel
+COPY --from=gen-channel /opt/ceremonyclient/target/release/libchannel.a /opt/ceremonyclient/target/release/libchannel.a
+
+COPY --from=gen-vdf /opt/ceremonyclient/vdf /opt/ceremonyclient/vdf
+COPY --from=gen-vdf /opt/ceremonyclient/target/release/libvdf.a /opt/ceremonyclient/target/release/libvdf.a
+
+COPY --from=gen-ferret /opt/ceremonyclient/ferret /opt/ceremonyclient/ferret
+COPY --from=gen-ferret /opt/ceremonyclient/target/release/libferret.a /opt/ceremonyclient/target/release/libferret.a
+
+COPY --from=gen-bls48581 /opt/ceremonyclient/bls48581 /opt/ceremonyclient/bls48581
+COPY --from=gen-bls48581 /opt/ceremonyclient/target/release/libbls48581.a /opt/ceremonyclient/target/release/libbls48581.a
+
+COPY --from=gen-rpm /opt/ceremonyclient/rpm /opt/ceremonyclient/rpm
+COPY --from=gen-rpm /opt/ceremonyclient/target/release/librpm.a /opt/ceremonyclient/target/release/librpm.a
+
+COPY --from=gen-verenc /opt/ceremonyclient/verenc /opt/ceremonyclient/verenc
+COPY --from=gen-verenc /opt/ceremonyclient/target/release/libverenc.a /opt/ceremonyclient/target/release/libverenc.a
+
+COPY --from=gen-bulletproofs /opt/ceremonyclient/bulletproofs /opt/ceremonyclient/bulletproofs
+COPY --from=gen-bulletproofs /opt/ceremonyclient/target/release/libbulletproofs.a /opt/ceremonyclient/target/release/libbulletproofs.a
+
+# -----------------------------------------------------------------------------
+# Stage: build-node
+# -----------------------------------------------------------------------------
+FROM build-context AS build-node
 COPY ./node /opt/ceremonyclient/node
 WORKDIR /opt/ceremonyclient/node
-
-ENV GOPROXY=direct
 RUN ./build.sh && cp node /usr/bin
 
-FROM build AS build-qclient
+# -----------------------------------------------------------------------------
+# Stage: build-qclient
+# -----------------------------------------------------------------------------
+FROM build-context AS build-qclient
 ARG TARGETOS
 ARG TARGETARCH
-# Build and install qclient
+
 COPY ./node /opt/ceremonyclient/node
-
 WORKDIR /opt/ceremonyclient/node
-
 RUN go mod download
 
 COPY ./client /opt/ceremonyclient/client
 WORKDIR /opt/ceremonyclient/client
-
 RUN go mod download
 
 ARG BINARIES_DIR=/opt/ceremonyclient/target/release
 RUN GOOS=${TARGETOS} GOARCH=${TARGETARCH} ./build.sh -o qclient
 RUN cp qclient /usr/bin
 
-# Allows exporting single binary
+# -----------------------------------------------------------------------------
+# Stage: node-only
+# -----------------------------------------------------------------------------
+FROM ubuntu:24.04 AS node-only
+ARG NODE_VERSION
+ARG GIT_REPO
+ARG GIT_BRANCH
+ARG GIT_COMMIT
+
+ENV GOEXPERIMENT=arenas
+
+LABEL org.opencontainers.image.title="Quilibrium Network Node"
+LABEL org.opencontainers.image.description="Quilibrium is a decentralized alternative to platform as a service providers."
+LABEL org.opencontainers.image.version=$NODE_VERSION
+LABEL org.opencontainers.image.vendor=Quilibrium
+LABEL org.opencontainers.image.url=https://quilibrium.com/
+LABEL org.opencontainers.image.documentation=https://quilibrium.com/docs
+LABEL org.opencontainers.image.source=$GIT_REPO
+LABEL org.opencontainers.image.ref.name=$GIT_BRANCH
+LABEL org.opencontainers.image.revision=$GIT_COMMIT
+
+RUN apt-get update && apt-get install -y \
+    ca-certificates \
+    libgmp10 \
+    libmpfr6 \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY --from=build-node /usr/bin/node /usr/local/bin/node
+WORKDIR /root
+ENTRYPOINT ["node"]
+
+# -----------------------------------------------------------------------------
+# Stage: exports (Scratch)
+# -----------------------------------------------------------------------------
 FROM scratch AS node
 COPY --from=build-node /usr/bin/node /node
 ENTRYPOINT [ "/node" ]
 
-# Allows exporting single binary
 FROM scratch AS qclient-unix
 COPY --from=build-qclient /usr/bin/qclient /qclient
 ENTRYPOINT [ "/qclient" ]
@@ -184,9 +281,15 @@ FROM qclient-unix AS qclient-linux
 FROM qclient-unix AS qclient-darwin
 FROM qclient-${TARGETOS} AS qclient
 
-FROM ubuntu:24.04
+# -----------------------------------------------------------------------------
+# Stage: final (Default combined image)
+# -----------------------------------------------------------------------------
+FROM ubuntu:24.04 AS final
 
-RUN apt-get update && apt-get install libflint-dev -y
+RUN apt-get update && apt-get install -y \
+    libflint-dev \
+    ca-certificates \
+    && rm -rf /var/lib/apt/lists/*
 
 ARG NODE_VERSION
 ARG GIT_REPO
@@ -205,11 +308,8 @@ LABEL org.opencontainers.image.source=$GIT_REPO
 LABEL org.opencontainers.image.ref.name=$GIT_BRANCH
 LABEL org.opencontainers.image.revision=$GIT_COMMIT
 
-RUN apt-get update && apt-get install -y ca-certificates
-
 COPY --from=build-node /usr/bin/node /usr/local/bin
 COPY --from=build-qclient /opt/ceremonyclient/client/qclient /usr/local/bin
 
 WORKDIR /root
-
 ENTRYPOINT ["node"]

